### PR TITLE
Onboarding Refactor | Bye react-albus :(

### PR DESCRIPTION
**Pull request by tbezman**
image

The changes in this PR aren't super significant, a lot of the changes are basically moving files around and duplicating a few pages, I'll elaborate on this duplication later.

### One Route => Many Routes
This change is a bit cosmetic but breaks us out of a god component and lets us split things up a bit. The following benefits are perceived

- Your step is saved when the user hits refresh
- Data is loaded more granularly (exactly what we need for the step).

### Duplication of logic
There are now three pages related to "Editing Collections". 
- Onboarding create collection
- Onboarding edit collection
- Logged in user edit collection

Instead of having a single page to edit a collection, we have three to make it much easier to reason about the logic for editing a collection. It's now obvious exactly what's going to happen when you click "Save" based on the file you're in.

### Colocation of `WizardFooter`
Now, each step renders their own footer
- Every step can completely control their own footer with idiomatic React.
- No having to reason about "Cleanup the global state when a component unmounts"
- Can do some really cool things with suspense and form states when we get React 18.

### Tests
- [x] Clicking "edit collection" from a gallery should bring you to the specific collection's editor
- [x] Clicking "edit collection" from a collection should bring you to the specific collection's editor
- [x] Entering collection edit from a gallery and clicking back should bring you back to the gallery
- [x] Entering collection edit from a gallery and clicking save should bring you back to the gallery (fixable with returnTo param)
- [x] Entering collection edit from a collection and clicking back should bring you back to the collection
- [x] Entering collection edit from a collection and clicking save should bring you back to the collection (fixable with returnTo param)
- [x] Clicking "cancel" from the editor should trigger confirmation modal
- [x] Clicking "cancel" button from collection organizer should return to where you were previously, defaulting to the user's profile
- [x] go back and forth in onboarding flow (probably more sub-tests to write here)

Other notes
- UI feels like it hangs for a bit as we move from step to step. Probably because we're downloading different bundles.

### CODE REVIEW

File: pages/edit.tsx [hunk 1]
```diff
@@ -1,25 +1,40 @@
-import GalleryAuthenticatedRoute from 'scenes/_Router/GalleryAuthenticatedRoute';
-import EditGalleryFlow from 'flows/EditGalleryFlow/EditGalleryFlow';
+/**
+ * This file exists strictly for backward compatability.
+ *
+ * Users that used to visit /edit directly will now be redirected
+ * to `/gallery/{firstGalleryId}/edit`.
```
Comment (Robinnnnn) [inline pages/edit.tsx:5]: love this

File: pages/gallery/[galleryId]/collection/[collectionId]/edit.tsx [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+  return (
+    <CollectionWizardContext initialCollectionId={collectionId}>
+      <CollectionEditorProvider>
+        <LazyLoadedCollectionEditor collectionId={collectionId} galleryId={galleryId} />
+      </CollectionEditorProvider>
+    </CollectionWizardContext>
+  );
+}
+
+export const getServerSideProps: GetServerSideProps<Props> = async ({ params }) => {
```
Comment (Robinnnnn) [inline pages/gallery/[galleryId]/collection/[collectionId]/edit.tsx:101]: the new directory layout is super nice

File: pages/gallery/[galleryId]/collection/create.tsx [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+  useCollectionMetadataState,
+  useStagedCollectionState,
+} from 'contexts/collectionEditor/CollectionEditorContext';
+import { useModalActions } from 'contexts/modal/ModalContext';
+import CollectionWizardContext from 'contexts/wizard/CollectionWizardContext';
+import { useTrack } from 'contexts/analytics/AnalyticsContext';
+import { graphql, useLazyLoadQuery } from 'react-relay';
+import CollectionCreateOrEditForm from 'flows/../../src/components/ManageGallery/OrganizeCollection/CollectionCreateOrEditForm';
+import CollectionEditor from 'flows/../../src/components/ManageGallery/OrganizeCollection/Editor/CollectionEditor';
+import FullPageCenteredStep from 'flows/../../src/components/Onboarding/FullPageCenteredStep/FullPageCenteredStep';
```
Comment (Robinnnnn) [inline pages/gallery/[galleryId]/collection/create.tsx:12]: these filepaths lol 🤯

File: pages/onboarding/congratulations.tsx [hunk 4]
```diff
@@ -1,24 +1,33 @@
...
-    throw new Error('Congratulations.tsx could not access queryRef');
-  }
-
-  const query = usePreloadedQuery(organizeCollectionQuery, queryRef);
-
-  if (query.viewer.__typename !== 'Viewer') {
+  if (query.viewer?.__typename !== 'Viewer') {
     throw new Error(
-      `OrganizeCollection expected Viewer to be type 'Viewer' but got: ${query.viewer.__typename}`
+      `OrganizeCollection expected Viewer to be type 'Viewer' but got: ${query.viewer?.__typename}`
```
Comment (Robinnnnn) [inline pages/onboarding/congratulations.tsx:30]: nit: error message should probably say `Congratulations expected Viewer ...`

File: pages/onboarding/create.tsx [hunk 5]
```diff
@@ -1,34 +1,26 @@
...
+  const { push, query, back } = useRouter();
   const handleNextClick = useCallback(() => {
     track('Start new collection');
-    next();
-  }, [next, track]);
 
-  useWizardConfig({ onNext: handleNextClick });
+    push({
+      pathname: getStepUrl('organize-collection'),
+      query: { ...query },
```
Comment (Robinnnnn) [inline pages/onboarding/create.tsx:21]: any reason not to just do `query: query`?

File: pages/onboarding/organize-collection.tsx [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+  const handleNext = useCallback(() => {
+    track('Save new collection button clicked');
+
+    showModal({
+      content: (
+        <CollectionCreateOrEditForm
+          onNext={() => {
+            push({
+              pathname: getStepUrl('organize-gallery'),
+              query: { ...urlQuery },
```
Comment (Robinnnnn) [inline pages/onboarding/organize-collection.tsx:70]: same comment - why do we need to make a copy?

File: src/contexts/SyncTokensLockContext.tsx [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+  lock: () => UnlockFunction;
+};
+
+const SyncTokensLockContext = createContext<SyncTokensLockContextType | undefined>(undefined);
+
+type Props = {
+  children: ReactNode;
+};
+
+export function SyncTokensLockProvider({ children }: Props) {
```
Comment (Robinnnnn) [inline src/contexts/SyncTokensLockContext.tsx:16]: SO ELEGANT

File: pages/gallery/[galleryId]/collection/create.tsx (hunk 3 above)
Comment (tbezman) [reply pages/gallery/[galleryId]/collection/create.tsx:12]: hmmm, is there some eslint plugin to make this shit nicer? this is so ugly but just what intellij did lol

File: pages/onboarding/congratulations.tsx (hunk 4 above)
Comment (tbezman) [reply pages/onboarding/congratulations.tsx:30]: true

File: pages/onboarding/create.tsx (hunk 5 above)
Comment (tbezman) [reply pages/onboarding/create.tsx:21]: nothing but paranoia lol

File: pages/gallery/[galleryId]/collection/create.tsx (hunk 3 above)
Comment (tbezman) [reply pages/gallery/[galleryId]/collection/create.tsx]: actually those paths were fucked and incorrect. intellij goofed during a file move refactor.

**Comment by kaitoo1** [comment]
> One Route => Many Routes

> There are now three pages related to "Editing Collections".

> Now, each step renders their own footer

man each one of these is so exciting

**Comment by kaitoo1** [comment]
clicked around to try and help test+ identify and edge cases.

refreshing worked like a charm, tried adding/deleting a bunch of collections and did some weird stuff I didn't notice any major issues 🔥

I did get stuck in one place, which I'd say is worth fixing:
- on /add-user-info, set a username and click "Next"
- on /create, click "Back"
- the username input is empty, and "next" is disabled so I need to enter something. But I can't continue with the same username (says it's taken), and I can't set a new username (since I've already created a user). so I'm stuck on the page unless I know to hit the browser back button.

screencap:

https://user-images.githubusercontent.com/80802871/195035925-ac026b43-c583-4a04-a4fb-55d4ff204be1.mp4


If the user returns to /add-user-info after setting their username and creating the user, can we treat it as editing an existing user and populate those fields?

**Comment by tbezman** [comment]
@kaitoo1 fixed the user editing thing. actually worked out pretty well.

**Comment by kaitoo1** [comment]
sick that was fast! just checked it out locally, looks great thanks @tbezman

**Pull request merged**